# Exercise 4.5

In [79]:
#import libraries
import pandas as pd
import numpy as np
import os

In [80]:
#assign path of the dataframes to a variable 'path'
path=r'C:\Users\19452\Desktop\Seema\Careerfoundry-Data Analytics Course\Work Area\Python Fundamentals\Nov-2023 Instacart Basket Analysis\02 Data'

# import products dataframe from original folder
df_prods=pd.read_csv(os.path.join(path,'Original Data','products.csv'),index_col=False)

# import orders_wrangled dataframe from prepared folder
df_ords=pd.read_csv(os.path.join(path,'Prepared Data','orders_wrangled.csv'),index_col=False)

## Step 1: Consistency checks

## Address Mixed data-types

In [81]:
# create a dataframe

df_test=pd.DataFrame()

# create a mixed type column

df_test['mix']=['a','b',1, True]

df_test.head()

mix
0     a
1     b
2     1
3  True

In [82]:
# The function for checking whether a dataframe contains any mixed-type columns is as follows:
# weird= (df_test[[col]].applymap(type)!=df_test[[col]].iloc[0].apply(type)).any(axis=1) -- This line gives error, 
# so changed applymap to map

for col in df_test.columns.tolist():
    weird= (df_test[[col]].map(type)!=df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_test[weird])>0:
        print (col)
    

mix


In [83]:
# this will convert a column’s data type from numeric to string

df_test['mix']=df_test['mix'].astype('str')

# Missing Values

In [84]:
# Find missing values in products dataframe
# The output shows there are 16 missing values in product_name colum

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [85]:
# to view 16 observations with missing values, you can create a subset of the dataframe 
# containing only the values in question

df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [86]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [87]:
# df.describe()

# df['column with missings'].fillna(mean value, inplace=True)

# df_prods.median()

# df['column with missings'].fillna(median value, inplace=True)


In [88]:
# Check the number of rows and columns before clean up
df_prods.shape

(49693, 5)

In [89]:
# extract only the observations from df_prods where product_name column is not blank and store it in a new dataframe
# df_prods_clean

df_prods_clean= df_prods[df_prods['product_name'].isnull() == False]

In [90]:
# Check the number of rows and columns in the new dataframe with extracted records after clean up

df_prods_clean.shape

(49677, 5)

In [91]:
#  drop all missing values from df_prods 

# df_prods.dropna(inplace=True)

# If you wanted to drop only the NaNs from a particular column.  Inplace=True with replace the existing dataframe with new.

# df_prods.dropna(subset='product_name',inplace=True)

In [92]:
# extract duplicate observations from df_prods_clean and store it in dataframe df_dups

df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [93]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [94]:
df_prods_clean.shape

(49677, 5)

In [95]:
# remove duplicates from df_prods_clean dataframe and store the results in df_prods_clean_no_dups

df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [96]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5

### Step 2
#### Based on the descriptive statistics I don't see any discrepancy with the data

In [97]:
#calculate descriptive statistics for each numeric column

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [98]:
#calculate descriptive statistics for each numeric column and suppress scientific notation

df_ords.describe().apply(lambda x: x.apply('{0:.5f}'.format))


Unnamed: 0       order_id        user_id   order_number  \
count  3421083.00000  3421083.00000  3421083.00000  3421083.00000   
mean   1710541.00000  1710542.00000   102978.20806       17.15486   
std     987581.73983   987581.73982    59533.71779       17.73316   
min          0.00000        1.00000        1.00000        1.00000   
25%     855270.50000   855271.50000    51394.00000        5.00000   
50%    1710541.00000  1710542.00000   102689.00000       11.00000   
75%    2565811.50000  2565812.50000   154385.00000       23.00000   
max    3421082.00000  3421083.00000   206209.00000      100.00000   

      orders_day_of_week order_hour_of_day days_since_prior_order  
count      3421083.00000     3421083.00000          3214874.00000  
mean             2.77622          13.45202               11.11484  
std              2.04683           4.22609                9.20674  
min              0.00000           0.00000                0.00000  
25%              1.00000          10.00000                4.00000  
50%              3.00000          13.00000                7.00000  
75%              5.00000          16.00000               15.00000  
max              6.00000          23.00000               30.00000

### Step 3
#### Check for mixed data-type columns

In [99]:
# Check whether df_ords contain mixed-type column

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
else :
    print ("None of the columns have mix data")

None of the columns have mix data


### Step 4
#### No mixed data-type columns found

### Step 5 
#### days_since_prior_order column has 206209 missing values

In [100]:
# Check if there are any missing values in df_ords dataframe

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### Step 6 
#### 206209 observations with missing values is a significant volume of data.  It's not safe to delete or impute them as this could lead to wrong analysis.  The dataframe df_ords_missing_values  has null values in all the observations with order_number=1 for every user.  This could be the first order of every user and it's logical not to have any value in days_since_prior_order column.  Also, there are 206209 users, which means every user will have one NaN value for his/her first order.  Considering all these factors it has been decided to leave the missing data untouched.

In [101]:
# df_ords dataframe has 3421083 observations including null values in days_since_prior_order

df_ords.shape

(3421083, 7)

In [102]:
## extract all the missing value observations to a new dataframe

df_ords_missing_values=df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [103]:
df_ords_missing_values

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

### Step 7
#### No duplicates were found in df_ords dataframe

In [104]:
# check for duplicate values in df_ords using duplicated function and assign the results to df_ords_diplicates dataframe.

df_ords_duplicates=df_ords[df_ords.duplicated()]

In [105]:
# check for duplicates in df_ords_duplicates

df_ords_duplicates

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### Step 8
#### An empty df_ords_duplicates dataframe proves there are no duplicates.  No action needed to remove duplicates.

### Step 9 

In [106]:
# export df_prods_clean_no_dups to prepared data folder as a new file products_checked.csv

df_prods_clean_no_dups.to_csv(os.path.join(path,'Prepared Data', 'products_checked.csv'))

# export df_ords to prepared data folder as a new file products_checked.csv

df_ords.to_csv(os.path.join(path,'Prepared Data', 'orders_checked.csv'))